src/categorical_algebra/DPO.jl

```diff
@@ -0,0 +1,215 @@
+
+    # Adjust lg function from I→G to refer to K using the offset
+    newFunc = [x - offsets[comp][x] for x in lm.components[comp].func]
+    k_components[comp] = FinFunction(newFunc, n_comp - length(orphans))
+  end
+
+  # Populate attributes for K (unchanged from G)
+  comps, arrows, src, tgt = typeof(G).parameters[1].parameters
+  attrs, srcs = typeof(G).parameters[2].parameters[3:4]
+  for (i, attr) in enumerate(attrs)
+    new=[G[attr][j] for j in g_components[comps[srcs[i]]]]
+    set_subpart!(K, attr, new)
+  end
+
+  # Populate relations for K (same as G, with offset applied)
+  for (i, col) in enumerate(arrows)
+    src_, tgt_ = comps[src[i]], comps[tgt[i]]
+    new=[val - offsets[tgt_][val] for val in G[col][g_components[src_]]]
+    set_subpart!(K, col, new)
+  end
+
+  # Put together all information into new morphisms
+  k = ACSetTransformation(I, K; k_components...)
+  @assert is_natural(k)
+  g = ACSetTransformation(K, G; g_components...)
+  @assert is_natural(g)
+
+  return k => g
+end
+
+"""
+Compute offsets induced by removing particular elements from
+the list 1:n. E.g. if we remove elements 2 and 4 from 1:5,
+our offset is [0,_,1,_,2]
+(the offset at the removed element index is not meaningful)
+"""
+function reindex(n::Int, removed::Vector{Int})::Vector{Int}
+  offset, off_counter, o_index = [],0,1
+  for i in 1:n
+    while o_index <= length(removed) && removed[o_index] < i
+      if removed[o_index] < i
+        off_counter +=1
+      end
+      o_index += 1
+    end
+    push!(offset, off_counter)
+  end
+  return offset
+end
+
+"""
+Rewrite with explicit match
+"""
+function rewrite_match(L::ACSetTransformation, R::ACSetTransformation, m::ACSetTransformation)::AbstractACSet
+    @assert L.dom == R.dom
+    @assert L.codom == m.dom
+    @assert valid_dpo(L, m)
+    @assert is_natural(L)
+    @assert is_natural(R)
+    (k, _) = pushout_complement(L, m)
+    l1, _ = pushout(R, k).cocone.legs
+    return l1.codom
+end
+
+"""
+Don't explicitly choose the match
+"""
+function rewrite(L::ACSetTransformation,
+                 R::ACSetTransformation,
+                 G::AbstractACSet,
+                 monic::Bool=false,
+                 m_index::Int=1
+                 )::Union{Nothing, AbstractACSet}
+  ms = filter(m->valid_dpo(L, m), homomorphisms(L.codom, G, monic=monic))
+  if 0 < m_index <= length(ms)
+    return rewrite_match(L, R, ms[m_index])
+  else
+    return nothing
+  end
+end
+
+
+"""
+Condition for existence of a pushout complement
+"""
+function valid_dpo(L::ACSetTransformation, m::ACSetTransformation, verbose::Bool=false)::Bool
+  return id_condition(L, m) && dangling_condition(L, m)
+end
+
+"""
+Does not map both a deleted item and a preserved item in L to the same item in G, or two distinct deleted items to the same.
+(Trivially satisfied if match is mono)
+"""
+function id_condition(L::ACSetTransformation, m::ACSetTransformation, verbose::Bool=false)::Bool
```

Comment:
`verbose` should be a keyword-only argument (not positional), so this would read `id(..., m::ACSetTransformation; verbose::Bool=false)` (note the semicolon).